File: cs/src/DataCentric/Types/Record/Record.cs
/*
Copyright (C) 2013-present The DataCentric Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

using System;
using CsvHelper.Configuration.Attributes;
using MongoDB.Bson.Serialization.Attributes;

namespace DataCentric
{
    /// <summary>
    /// Base class of records stored in data source.
    /// </summary>
    [BsonDiscriminator(RootClass = true)]
    public abstract class Record : Data
    {
        /// <summary>
        /// Execution context provides access to key resources including:
        ///
        /// * Logging and error reporting
        /// * Cloud calculation service
        /// * Data sources
        /// * Filesystem
        /// * Progress reporting
        /// </summary>
        [BsonIgnore]
        [Ignore]
        public IContext Context { get; private set; }

        //--- ELEMENTS

        /// <summary>
        /// RecordId of the record is specific to its version.
        ///
        /// For the record's history to be captured correctly, all
        /// update operations must assign a new RecordId with the
        /// timestamp that matches update time.
        /// </summary>
        [BsonId]
        [BsonRequired]
        public RecordId Id { get; set; }

        /// <summary>
        /// RecordId of the dataset where the record is stored.
        ///
        /// For records stored in root dataset, the value of
        /// DataSet element should be RecordId.Empty.
        /// </summary>
        [BsonElement("_dataset")]
        public RecordId DataSet { get; set; }

        /// <summary>
        /// String key consists of semicolon delimited primary key elements:
        ///
        /// KeyElement1;KeyElement2
        ///
        /// To avoid serialization format uncertainty, key elements
        /// can have any atomic type except Double.
        /// </summary>
        [BsonElement("_key")]
        [BsonRequired]
        public abstract string Key { get; set; }

        //--- METHODS

        /// <summary>
        /// Set Context property and perform validation of the record's data,
        /// then initialize any fields or properties that depend on that data.
        ///
        /// This method may be called multiple times for the same instance,
        /// possibly with a different context parameter for each subsequent call.
        ///
        /// IMPORTANT - Every override of this method must call base.Init()
        /// first, and only then execute the rest of the override method's code.
        /// </summary>
        public virtual void Init(IContext context)
        {
            // Uncomment except in root class of the hierarchy
            // base.Init(context);

            // Check that argument is not null and assign to the Context property
            if (context == null) throw new Exception($"Null context is passed to the Init(...) method for {GetType().Name}.");
            Context = context;
        }

        /// <summary>Return string representation of the record's key.</summary>
        public override string ToString() { return Key; }
    }

    /// <summary>Extension methods for Record.</summary>
    public static class RecordExtensions
    {
        /// <summary>Deserialize record from XML using short
        /// class name without namespace for the root XML element.</summary>
        public static void ParseXml(this Record obj, string xmlString)
        {
            ITreeReader reader = new XmlTreeReader(xmlString);

            // Root node of serialized XML must be the same as mapped class name without namespace
            var mappedFullName = obj.GetType().Name;
            ITreeReader recordNodes = reader.ReadElement(mappedFullName);

            // Deserialize from XML nodes inside the root node
            obj.DeserializeFrom(recordNodes);
        }

        /// <summary>Serialize record to XML using short
        /// class name without namespace for the root XML element.</summary>
        public static string ToXml(this Record obj)
        {
            // Get root XML element name using mapped final type of the object
            string rootName = obj.GetType().Name;

            // Serialize to XML
            ITreeWriter writer = new XmlTreeWriter();
            writer.WriteStartDocument(rootName);
            obj.SerializeTo(writer);
            writer.WriteEndDocument(rootName);

            // Convert to string
            string result = writer.ToString();
            return result;
        }
    }
}

File: py/src/datacentric/types/record/record.py
from abc import ABC, abstractmethod
from bson import ObjectId

from datacentric.platform.context import Context
from datacentric.types.record import Data


class Record(Data, ABC):
    """Base class of records stored in data source.
    """
    __slots__ = ['context', 'id_', 'data_set', '_key']

    context: Context
    id_: ObjectId
    data_set: ObjectId
    _key: str

    def __init__(self):
        super().__init__()

        self.context = None
        """Execution context"""

        self.id_ = None
        """ObjectId of the record is specific to its version.
        For the record's history to be captured correctly, all
        update operations must assign a new ObjectId with the
        timestamp that matches update time.
        """

        self.data_set = None
        """ObjectId of the dataset where the record is stored.
        For records stored in root dataset, the value of
        data_set element should be ObjectId('000000000000000000000000').
        """

        self._key = None
        """Backing attribute for key() property."""

    def init(self, context: Context) -> None:
        if context is None:
            raise Exception(f'Null context is passed to the Init(...) method for {type(self).__name__}."')
        self.context = context

    @property
    @abstractmethod
    def key(self) -> str:
        """String key consists of semicolon delimited primary key elements:

        key_element1;key_element2

        To avoid serialization format uncertainty, key elements
        can have any atomic type except float.
        """
        pass
